<launch>

  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="open_rviz" default="true"/>
  <arg name="rtabmapviz" default="true"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="map_output_topic" default="/map"/>

  <arg name="with_camera" default="false"/>
  <arg name="localization" default="false"/>
  <arg name="database_path" default="~/.ros/rtabmap.db"/>
  <arg if="$(arg localization)" name="rtabmap_args" default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args" default="--delete_db_on_start"/>

  <group ns="rtabmap">
    <node if="$(eval model=='waffle')" pkg="rtabmap_ros" type="rgbd_sync" name="rgbd_sync" output="screen">
      <remap from="rgb/image" to="/camera/rgb/image_raw"/>
      <remap from="depth/image" to="/camera/depth/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
    </node>
  
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <param name="database_path" type="string" value="$(arg database_path)"/>
      <param name="frame_id" type="string" value="base_footprint"/>
      <param name="subscribe_rgb" type="bool" value="false"/>
      <param name="subscribe_depth" type="bool" value="false"/>
      <param if="$(eval model=='waffle')" name="subscribe_rgbd" type="bool" value="true"/>
      <param unless="$(eval model=='waffle')" name="subscribe_rgbd" type="bool" value="false"/>
      <param name="subscribe_scan" type="bool" value="true"/>
      <param name="approx_sync" type="bool" value="true"/>
	
      <!-- inputs -->
      <remap from="scan" to="/scan"/>
      <remap from="odom" to="/odom"/>
      <remap from="rgbd_image" to="rgbd_image"/>
      
      <!-- output -->
      <remap from="grid_map" to="$(arg map_output_topic)"/>
	
      <!-- RTAB-Map's parameters -->
      <param name="Reg/Strategy" type="string" value="1"/>
      <param name="Reg/Force3DoF" type="string" value="true"/>
      <param name="GridGlobal/MinSize" type="string" value="10"/>
      <param name="map_always_update" type="bool" value="true"/>
      <param name="map_empty_ray_tracing" type="bool" value="true"/>
      <param name="map_filter_radius" type="double" value="0.1"/>
      <param name="map_filter_angle" type="double" value="30"/>
      <param name="grid_size" type="double" value="5"/> 
      
      <!-- localization mode -->
      <param if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
    </node>

    <!-- visualization with rtabmapviz -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_scan" type="bool" value="true"/>
      <param name="subscribe_odom" type="bool" value="true"/>
      <param name="frame_id" type="string" value="base_footprint"/>
      <param name="approx_sync" type="bool" value="true"/>

      <remap from="odom" to="/odom"/>
      <remap from="scan" to="/scan"/>
    </node>
  </group>

</launch>
